script/openqa-powermanagement.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/python3
+
+# TODO:
+#  * Add return values checks to avoid crashes/traces (no connection to openQA server, etc.)
+#  * The host name of machines may not be unique
+
+import configparser
+import argparse
+import json
+import requests
+import subprocess
+
+machine_list_idle = []
+machine_list_offline = []
+machine_list_busy= []
+machines_to_power_on = []
+
+jobs_worker_classes = []
+
+config_file = "/etc/openqa/openqa.ini"
+config = configparser.ConfigParser()
+config.read(config_file)
+
+openqa_server = "http://localhost"
+
+# Manage cmdline options
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--config')
+    parser.add_argument('--host')
+    parser.add_argument('--osd', action='store_true')
+    parser.add_argument('--o3', action='store_true')
+    args = parser.parse_args()
+    if args.config is not None and len(args.config):
+      config_file = args.config
+    if args.host is not None and len(args.host):
+      openqa_server = args.host
+    elif args.osd:
+      openqa_server = "http://openqa.suse.de"
+    elif args.o3:
+      openqa_server = "https://openqa.opensuse.org"
+
+print("Using openQA server: " + openqa_server)
+print("Using config file: " + config_file)
+print("")
+
+# Scheduled/blocked jobs
+scheduled_list_file = requests.get(openqa_server + '/tests/list_scheduled_ajax').content
+scheduled_list_data = json.loads(scheduled_list_file)
+print("Processing " + str(len(scheduled_list_data['data'])) + " job(s) in scheduled/blocked state... (will take about " + str(int(len(scheduled_list_data['data']) * 0.2)) + " seconds)")
+  
+# Create list of WORKER_CLASS needed
+for job in scheduled_list_data['data']:
+  response = requests.get(openqa_server + '/api/v1/jobs/' + str(job['id']))
+  job_data = json.loads(response.content)
+  jobs_worker_classes.append(job_data['job']['settings']['WORKER_CLASS'])
+
+jobs_worker_classes = sorted(set(jobs_worker_classes))
+print("Found " + str(len(jobs_worker_classes)) + " different WORKER_CLASS in scheduled jobs: " + str(jobs_worker_classes))
+
+
+
+# Workers
+workers_list_file = requests.get(openqa_server + '/api/v1/workers').content
+workers_list_data = json.loads(workers_list_file)
+
+# Create list of hosts which may need to powered up/down
+for worker in workers_list_data['workers']:
+  if worker['status'] in ['idle']:
+    machine_list_idle.append(worker['host'])
+  elif worker['status'] in ['dead']: # Looks like 'dead' means 'offline'
```

Comment:
More or less. When I read the code correctly, you might not get `dead` here in case the worker quit normally. Then you might get `broken` instead with `graceful disconnect at …` as "error". That doesn't make very much sense but I suppose that's how it is now. (The stale worker/job detection grew historically.)

Reply:
From my tests, `dead` is used at least when the worker disconnected gracefully. I did not encountered the `broken` state yet.